homeassistant/components/ecoforest/coordinator.py

```diff
@@ -0,0 +1,49 @@
+"""The ecoforest coordinator."""
+
+
+import logging
+
+import async_timeout
+from pyecoforest.api import EcoforestApi
+from pyecoforest.exceptions import EcoforestAuthenticationRequired, EcoforestError
+from pyecoforest.models.device import Device
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import POLLING_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EcoforestCoordinator(DataUpdateCoordinator[Device]):
+    """DataUpdateCoordinator to gather data from ecoforest device."""
+
+    def __init__(self, hass: HomeAssistant, api: EcoforestApi, host: str) -> None:
+        """Initialize DataUpdateCoordinator."""
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="ecoforest",
+            update_interval=POLLING_INTERVAL,
+        )
+        self.api = api
+        self.host = host
+
+    async def _async_update_data(self) -> Device:
+        """Fetch all device and sensor data from api."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                data: Device = await self.api.get()
+                _LOGGER.debug("Ecoforest data: %s", data)
+                return data
+        except EcoforestAuthenticationRequired as err:
+            # Raising ConfigEntryAuthFailed will cancel future updates
+            # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+            raise ConfigEntryAuthFailed from err
```

Comment:
This comment is right as well, except for the fact, you don't have reauth yet. So best way is to add the initial data check in `__init__.py` like I commented before, so if the auth is invalid, we don't start the integration, but if something happened during a HA session we just keep looping (It's not nice but I think it's the best we can do)